fml/platform/fuchsia/message_loop_fuchsia.cc

```diff
@@ -6,33 +6,57 @@
 
 #include <lib/async-loop/default.h>
 #include <lib/async/cpp/task.h>
-#include <lib/zx/time.h>
 
 namespace fml {
 
 MessageLoopFuchsia::MessageLoopFuchsia()
-    : loop_(&kAsyncLoopConfigAttachToCurrentThread) {}
+    : loop_(&kAsyncLoopConfigAttachToCurrentThread), running_(false) {
+  auto status =
+      zx::timer::create(ZX_TIMER_SLACK_LATE, ZX_CLOCK_MONOTONIC, &timer_);
+  FML_CHECK(status == ZX_OK);
+  timer_wait_ = std::make_unique<async::Wait>(timer_.get(), ZX_TIMER_SIGNALED);
+
+  auto wait_handler = [this](async_dispatcher_t* dispatcher, async::Wait* wait,
+                             zx_status_t status,
+                             const zx_packet_signal_t* signal) {
+    timer_.cancel();
```

Comment:
Please perform the `ZX_ERR_CANCELED` check before you you do anything else that may access the captured `this`. When the message loop is terminated, the cancellation callback may occur after the `MessageLoopFuchsia` is collected via the dtor. When this happens `this->timer_` and `this->RunExpiredTasksNow()` will deference garbage data. Just return early if the status is cancelled. Message loop termination will already flush or drop pending tasks.

Another way to make certain that that you are not accessing the timer and task set on an expired message loop is to use a `WeakPtrFactory` to the message loop and have that vend a `WeakPtr` that is used in the capture instead of the raw `this` pointer. This isn't strictly necessary but may be a clearer indication about your intent around safe use of a potentially collected `this` capture.